app/src/org/commcare/android/nfc/NfcReadActivity.java

```diff
@@ -0,0 +1,119 @@
+package org.commcare.android.nfc;
+
+import android.annotation.TargetApi;
+import android.content.Intent;
+import android.nfc.FormatException;
+import android.nfc.NdefMessage;
+import android.nfc.NdefRecord;
+import android.nfc.Tag;
+import android.nfc.tech.Ndef;
+import android.os.Build;
+import android.os.Bundle;
+import android.util.Pair;
+
+import org.commcare.android.javarosa.IntentCallout;
+
+import java.io.IOException;
+
+/**
+ * When this activity is in the foreground, any NFC tag scanned by the device that is of a format
+ * CommCare recognizes (as determined by the filters in NfcActivity::setReadyToHandleTag) will be
+ * consumed by this activity. If the tag is of the specified/expected type(s), then this activity
+ * will attempt to read the tag. If the read is successful, the value read will be stored as an
+ * extra in the activity's result intent under the key "odk_intent_data", where it can be accessed
+ * by other activities. Otherwise, an appropriate error toast will be shown.
+ *
+ * @author Aliza Stone
+ */
+@TargetApi(Build.VERSION_CODES.JELLY_BEAN)
```

Comment:
Can this annotation be moved into the super class?

Reply:
oo yes good call, this got missed out when I did some refactoring

Reply:
Actually, I take this back, this should get removed entirely because nothing is actually using/respecting it anymore. I'm now taking care of this via the version code checks in `onCreate` and `onResume`

Reply:
Ok final update-- the correct thing here was to move these targets to the method in each class that actually does the work, so `readFromNfcTag()` and `writeMessageToNfcTag()`, since those are the ones that actually need to be protected from being called by lower versions (and are being protected by the check in `onResume`)